### ISSUE TITLE
BasePipelineTest.runScript(Script script) throws MissingMethodException

### ISSUE THREAD

**Issue by Willem1987** [jenkinsci/JenkinsPipelineUnit#269]
helper.runScript(script, this.binding) does not exists, should be helper.runScript(script)

groovy.lang.MissingMethodException: No signature of method: com.lesfurets.jenkins.unit.PipelineTestHelper.runScript() is applicable for argument types: (buildFromGitlab, groovy.lang.Binding) values: [buildFromGitlab@29852487, groovy.lang.Binding@771db12c]
Possible solutions: runScript(java.lang.String, groovy.lang.Binding), runScript(groovy.lang.Script), runScript(java.lang.String), loadScript(java.lang.String, groovy.lang.Binding)

	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:58)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:54)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
	at com.lesfurets.jenkins.unit.BasePipelineTest.runScript(BasePipelineTest.groovy:315)
	at org.codehaus.groovy.vmplugin.v7.IndyInterface.selectMethod(IndyInterface.java:232)
	at BuildTest.testBuild(BuildTest.groovy:33)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:599)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:822)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:147)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.testng.TestRunner.privateRun(TestRunner.java:764)
	at org.testng.TestRunner.run(TestRunner.java:585)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:384)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:378)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:337)
	at org.testng.SuiteRunner.run(SuiteRunner.java:286)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1218)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
	at org.testng.TestNG.runSuites(TestNG.java:1069)
	at org.testng.TestNG.run(TestNG.java:1037)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:109)

**Comment by nre-ableton** [comment]
`PipelineTestHelper.runScript(String, Binding)` [does exist](https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/src/main/groovy/com/lesfurets/jenkins/unit/PipelineTestHelper.groovy#L412). It looks like you are attempting to pass a non-string object as the first argument (`buildFromGitlab`). I'm guessing you are trying to pass a class there instead of the name of the script?

**Comment by Willem1987** [comment]
It is not me that is trying to do this. There is 2 methods in BasePipelineTest. 

Object runScript(Script script) and Object runScript(String scriptName).

The option to run a previously loaded script seems to call the wrong helper method. https://github.com/jenkinsci/JenkinsPipelineUnit/pull/270

**Comment by nre-ableton** [comment]
The error message states, `runScript() is applicable for argument types: (buildFromGitlab, groovy.lang.Binding)`, so this is why I suspected that the call was incorrect. Can I please see your test code where you are trying to call `runScript`?

I am not saying that the code is perfect and bug-free, for what it's worth. 😄 But based on the error log, I don't yet think it's a problem with the call to `helper.runScript`.

**Comment by nre-ableton** [comment]
Ah, sorry, I think I mixed up `String` and `Script` visually. You are correct. I'll approve the PR.

**Comment by nre-ableton** [comment]
Fixed by https://github.com/jenkinsci/JenkinsPipelineUnit/pull/270.

**Comment by cwiggs** [comment]
I can into this issue today when I upgraded from 1.3 to 1.7.  I was passing in a previous loaded script object into the `runScript()`.  e.g.:

```
...
class techopsSharedLibsTest extends BasePipelineTest {
    def script

    @Rule public ExpectedException thrown = ExpectedException.none();

    @Before
    void setUp() {
        super.setUp()
        script = loadScript("Jenkinsfile")

        // We need to setup the scm variable or the test will fail to find it.
        binding.setVariable('scm', [:])

        // Example of a mock sh step with no expected output(closure).
        helper.registerAllowedMethod("sh", [String.class], null)
        helper.registerAllowedMethod("withEnv", [List, Closure], null)
    }

    @Test void 'Verify gradle test failure actually fails.'() throws Exception {
        helper.registerAllowedMethod("sh", [String.class], {cmd ->
            if (cmd.contains('./gradlew test')) {
                throw new Exception("error message")
            }
        })

        thrown.expect(Exception)
        thrown.expectMessage('error message')
        runScript(script)
    }
}
```

I think #270 will fix this issue, but it doesn't look like that change has been released yet (maybe in 1.8?).  The workaround I used was loading the script using `runScript("Jenkinsfile")` multiple times. Example:

```
class techopsSharedLibsTest extends BasePipelineTest {
    def script

    @Rule public ExpectedException thrown = ExpectedException.none();

    @Before
    void setUp() {
        super.setUp()

        // We need to setup the scm variable or the test will fail to find it.
        binding.setVariable('scm', [:])

        // Example of a mock sh step with no expected output(closure).
        helper.registerAllowedMethod("sh", [String.class], null)
        helper.registerAllowedMethod("withEnv", [List, Closure], null)
    }
    @Test void 'Verify gradle test failure actually fails.'() throws Exception {
        helper.registerAllowedMethod("sh", [String.class], {cmd ->
            if (cmd.contains('./gradlew test')) {
                throw new Exception("error message")
            }
        })

        thrown.expect(Exception)
        thrown.expectMessage('error message')
        runScript("Jenkinsfile")
    }
}
```

I hope that helps someone else.
